src/graphics/mesh3d.rs

```diff
@@ -0,0 +1,336 @@
+    where
+        V: Into<mint::Vector3<f32>>,
+    {
+        let pivot: mint::Vector3<f32> = center.into();
+        let transform =
+            Mat4::from_translation(Vec3::from(param.transform.position) + Vec3::from(pivot))
+                * Mat4::from_scale(param.transform.scale.into())
+                * Mat4::from_quat(param.transform.rotation.into())
+                * Mat4::from_translation(
+                    (Vec3::from(param.transform.position) + Vec3::from(pivot)) * -1.0,
+                );
+
+        Self {
+            transform: [
+                transform.x_axis.into(),
+                transform.y_axis.into(),
+                transform.z_axis.into(),
+                transform.w_axis.into(),
+            ],
+            color: param.color.into(),
+        }
+    }
+}
+
+// TODO: Allow custom vertex formats
+/// The 3d Vertex format. Used for constructing meshes. At the moment it supports color, position, and texture coords
+#[derive(Clone, Copy, bytemuck::Zeroable, bytemuck::Pod, Debug)]
+#[repr(C)]
+pub struct Vertex3d {
+    /// The position of this vertex
+    pub pos: [f32; 3],
+    /// The texture uv of this vertex
+    pub tex_coord: [f32; 2],
+    /// The color of this vertex
+    pub color: [f32; 4],
+}
+
+impl Vertex3d {
+    /// Create a new vertex from a position, uv, and color
+    pub fn new<V, T, C>(position: V, uv: T, color: C) -> Vertex3d
+    where
+        V: Into<Vector3<f32>>,
+        T: Into<Vector2<f32>>,
+        C: Into<Option<graphics::Color>>,
+    {
+        let position: Vector3<f32> = position.into();
+        let uv: Vector2<f32> = uv.into();
+        let color: Option<graphics::Color> = color.into();
+        let color = color
+            .unwrap_or(graphics::Color::new(1.0, 1.0, 1.0, 0.0))
+            .into();
+        Vertex3d {
+            pos: position.into(),
+            tex_coord: uv.into(),
+            color,
+        }
+    }
+
+    pub(crate) fn desc() -> wgpu::VertexBufferLayout<'static> {
+        wgpu::VertexBufferLayout {
+            array_stride: std::mem::size_of::<Vertex3d>() as _,
+            step_mode: wgpu::VertexStepMode::Vertex,
+            attributes: &[
+                // pos
+                wgpu::VertexAttribute {
+                    format: wgpu::VertexFormat::Float32x3,
+                    offset: 0,
+                    shader_location: 0,
+                },
+                // tex_coord
+                wgpu::VertexAttribute {
+                    format: wgpu::VertexFormat::Float32x2,
+                    offset: std::mem::size_of::<[f32; 3]>() as wgpu::BufferAddress,
+                    shader_location: 1,
+                },
+                //color
+                wgpu::VertexAttribute {
+                    format: wgpu::VertexFormat::Float32x4,
+                    offset: std::mem::size_of::<[f32; 5]>() as wgpu::BufferAddress,
+                    shader_location: 2,
+                },
+            ],
```

Comment:
You could use [`wgpu::vertex_attr_array!`](https://docs.rs/wgpu/latest/wgpu/macro.vertex_attr_array.html) here too

Reply:
Fixed in latest commit